### PR TITLE
syntax.rs introduction

### PR DESCRIPTION
**Pull request by nitishch**
Ported `syntax-table` and `set-syntax-table`.

C's `PER_BUFFER_VAR_IDX` gets offset of a field and then indexes `buffer_local_flags` with that offset. Replaced that with the field access in `buffer_local_flags`.

Unrelated changes:
- Previously, if a ported function A used `#[lisp_fn(name='lisp-name')]`, `describe-function` for A couldn't locate the function in the file. Fixed it.

### CODE REVIEW

File: rust_src/src/lib.rs [hunk 1]
```diff
@@ -60,6 +60,7 @@ mod editfns;
 mod eval;
 mod ffi;
 mod fileio;
+mod syntax;
```
Comment (shaleh) [inline rust_src/src/lib.rs]: Note the order that was here.

File: rust_src/src/buffers.rs [hunk 2]
```diff
@@ -309,6 +311,23 @@ impl LispBufferRef {
         self.zv = charpos;
         self.zv_byte = byte;
     }
+
+    #[inline]
+    pub fn set_syntax_table(&mut self, table: LispCharTableRef) {
+        self.syntax_table = LispObject::from(table).to_raw();
+    }
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Why is there no blank line next?

File: rust_src/src/remacs_sys.rs [hunk 3]
```diff
@@ -1692,7 +1692,14 @@ extern "C" {
     ) -> LispObject;
     pub fn buffer_local_value(variable: LispObject, buffer: LispObject) -> LispObject;
     pub fn downcase(c: c_int) -> c_int;
-
+    pub fn scan_lists(
+        from: EmacsInt,
+        count: EmacsInt,
+        depth: EmacsInt,
+        sexpflag: bool,
+    ) -> LispObject;
```
Comment (shaleh) [inline rust_src/src/remacs_sys.rs]: Only functions here please. Variables are further up.

File: rust_src/src/buffers.rs (hunk 2 above)
Comment (nitishch) [reply rust_src/src/buffers.rs]: Fixed.

File: rust_src/src/remacs_sys.rs (hunk 3 above)
Comment (nitishch) [reply rust_src/src/remacs_sys.rs]: Fixed.

File: rust_src/src/lib.rs (hunk 1 above)
Comment (nitishch) [reply rust_src/src/lib.rs]: Fixed.

**Comment by nitishch** [comment]
Appveyor build is failing a benchmark test. From https://github.com/Wilfred/remacs/pull/735, I guess this is not related to the branch's changes.

File: lisp/emacs-lisp/find-func.el [hunk 4]
```diff
@@ -297,13 +297,17 @@ TYPE should be nil to find a function, or `defvar' to find a variable."
...
+                   `(and "fn" (+ space) ,rust-fname "(")))
+           (lisp-fn-name-regex (rx-to-string
+                                `(and "#[" (* space) "lisp_fn" (* space)
+                                      "(" (* space) "name" (* space) "="
+                                      (* space) "\"" ,fname))))
       (with-current-buffer (find-file-noselect file)
         (goto-char (point-min))
-        (unless (re-search-forward regex nil t)
-          (error "Can't find source for %s" identifier))
+        (unless (re-search-forward lisp-fn-name-regex nil t)
```
Comment (nitishch) [inline lisp/emacs-lisp/find-func.el]: Give preference to declarations which specify name using `name` attribute to `lisp_fn`. This fixes #749 too :).

File: rust_src/src/syntax.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#[lisp_fn]
+pub fn syntax_table() -> LispObject {
+    LispObject::from_raw(ThreadState::current_buffer().syntax_table)
+}
+
+/// Scan from character number FROM by COUNT lists.
+/// Scan forward if COUNT is positive, backward if COUNT is negative.
+/// Return the character number of the position thus found.
+///
+/// A \"list", in this context, refers to a balanced parenthetical
```
[the remark this answers is not in the document]
Comment (nitishch) [reply rust_src/src/syntax.rs]: Done. Overlooked this before.

File: rust_src/src/buffers.rs [hunk 6]
```diff
@@ -309,6 +311,24 @@ impl LispBufferRef {
...
+
+    ///Set whether per-buffer variable with index IDX has a buffer-local
```
[the remark this answers is not in the document]
Comment (nitishch) [reply rust_src/src/buffers.rs]: Done.

File: rust_src/src/buffers.rs [hunk 7]
```diff
@@ -309,6 +311,24 @@ impl LispBufferRef {
...
+    }
+
+    /// Set whether per-buffer variable with index IDX has a buffer-local
+    /// value in buffer.  VAL zero means it does't.
+    // Similar to SET_PER_BUFFER_VALUE_P macro in C
+    #[inline]
+    pub fn set_per_buffer_value_p(&mut self, idx: usize, val: c_uchar) {
+        unsafe {
+            if idx >= last_per_buffer_idx {
+                emacs_abort()
```
Comment (brotzeit) [inline rust_src/src/buffers.rs]: I think we use `panic!()`

Comment (nitishch) [reply rust_src/src/buffers.rs]: Replaced. Sorry. I forgot about this from the Porting Tips

Comment (shaleh) [reply rust_src/src/buffers.rs]: The code is currently inconsistent. We should decide on which we prefer and do a quick cleanup.

**Pull request merged**